blog/2019-03-29-why-haskell.md

```diff
@@ -0,0 +1,108 @@
+
+Here is a chart of our AWS expenditure as we switched to a Haskell backend:
+
+<Insert chart>
+
+## The Bad
+
+### Lack of conventions
+
+> There are very few well-trodden paths in Haskell, expect to make lots of decisions about your stack.
+
+Apart from things like `Yesod`, there aren’t really many framework style Haskell projects that come with best practices on how to structure your application or guides on how to do common tasks like handling file uploads, user authentication, database management etc.
+
+Working on Haskell projects feels very much like working with `react` projects in that you have to bring in external libraries piecemeal for most of the features you want and the libraries aren’t necessarily designed with the assumption that they will be used together in that particular combination.
+
+Conventions also codify some hard-learned lessons, in their absence most of those lessons have to be re-learned by your team.
+
+### Lack of libraries
+
+> While the quality of libraries is excellent, the quantity can be rather limiting. Don’t expect the equivalent of passportjs or paperclip.
+
+The best Haskell libraries tend to be rather low-level, providing the building blocks rather than solving one cohesive problem by themselves.
+
+For instance, you won’t find any ‘batteries included’ libraries for user authentication which support OAuth integration, work with most major providers, have a password reset functionality, etc. If you are coming from other technologies like Rails, where creating admin interfaces is a simple one day job with `activeadmin`, you would be surprised at how cumbersome and repetitive some of the common tasks can get.  
+
+### First-class developer experience
+
+> Expect Haskell to get only community support or delayed support in most places.
+
+When AWS releases interesting new tools like Lambda, expect to not be able to play around with them through Haskell right away. You will be a second-class citizen in most developer products.
+
+This translates to either missed opportunities or just extra time spent trying to setup common tools like CI systems which are designed to work well with popular languages but have trouble adjusting to Haskell projects (I’m looking at you CircleCI).
+
+## The Ugly
+
+### Breaking changes in libraries
+
+> You can expect libraries to have breaking changes that have ripple effects throughout your project.
+
+Most Haskell projects don’t really have a concept of backwards compatibility. They regularly release breaking changes because they tend to be either hobby projects or projects that simply have ambitious technical goals which don’t necessarily align with your company’s goals of having stable interfaces.
+
+Upgrading to a new stack LTS almost always ends up becoming a blocking task that requires many tangential changes or reviewing changelogs carefully of all the libraries that you depend on to make sure that there aren’t any behaviour changes.
```

Comment:
Probably not to make it build-tool opinionated but to say upgrading to the newer GHC or just to the newer versions?